Enhance param handling with param type descriptors

levitte opened the pull request:
Some utility functions to get and set algorithm and operation params were missing, and functions returning param type descriptors were missing completely.

This changeset adds the framework for them, and enhances `openssl list` to be able to describe the parameters for implementations that the provider is willing to share.

levitte committed:
Add missing EVP param utility functions

These functions were missing for a completes API:

EVP_MD_get_params(), EVP_CIPHER_get_params(), EVP_CIPHER_CTX_set_params(),
and EVP_CIPHER_CTX_get_params

Additionally, we also add all the corresponding parameter descriptor
returning functions, along the correspoding provider dispatches:

EVP_MD_get_param_types(), EVP_MD_CTX_set_param_types(),
EVP_MD_CTX_get_param_types(), EVP_CIPHER_get_param_types(),
EVP_CIPHER_CTX_set_param_types(), and EVP_CIPHER_CTX_get_param_types()

levitte committed:
Modify 'openssl list' to give more verbose descriptions

With the diverse {get,set}_param_types functions, it's possible to
give a more verbose description of the diverse algorithm
implementations.  Most notably, we add a description of the parameters
that each implementation is willing to share.

levitte committed:
Add {get,set}_param_types() functions for provider digests

levitte commented:
This isn't complete.  More documentation needs to be added, at the very least.

paulidale commented on apps/list.c:
```diff
@@ -68,6 +167,11 @@ static void list_ciphers(void)
         BIO_printf(bio_out, " @ %s\n",
                    OSSL_PROVIDER_name(EVP_CIPHER_provider(c)));
+        if (verbose) {
+            print_param_types(GET, ALGO, EVP_CIPHER_get_param_types(c));
```
I think that this would this be clearer skipping the enum arguments and arrays and just passing a title:
```
print_param_types("retrievable algorithm", EVP_CIPER_get_param_types(c));
print_param_types("retrievable operation", EVP_CIPHER_CTX_get_param_types(c));
print_param_types("settable operation", EVP_CIPHER_CTX_set_param_types(c));
```

levitte replied:
Good point

paulidale replied:
The same sequence appears again later in the file, so it needs to be hit in two places.

levitte replied:
Done

paulidale replied:
Awaiting push?

levitte replied:
Oh...  done

paulidale reviewed (commented):
I really like this.

levitte commented:
> I really like this.

That makes me glad :-)

levitte committed:
fixup! Modify 'openssl list' to give more verbose descriptions

levitte committed:
fixup! Add missing EVP param utility functions

levitte committed:
fixup! Modify 'openssl list' to give more verbose descriptions

levitte committed:
fixup! Add missing EVP param utility functions

levitte committed:
fixup! Modify 'openssl list' to give more verbose descriptions

levitte commented:
Okie, out of WIP.  Time for more thorough reviews

mattcaswell commented on doc/man3/EVP_DigestInit.pod:
```diff
@@ -121,16 +128,29 @@ EVP_MD_CTX_ctrl() must be called after EVP_DigestInit_ex(). Other restrictions
+
+Get a B<OSSL_PARAM> array that describes the retrievable and settable
+parameters, i.e. parameters that can be used with EVP_MD_get_params(),
+EVP_MD_CTX_get_params() and EVP_MD_CTX_set_params(), respectively.
```
So this is kind of confusing. If I'm reading this correctly `EVP_MD_CTX_set_param_types` is actually a "getter" to get the param types for `EVP_MD_CTX_set_params`. That seems wrong somehow.

levitte replied:
The alternative is names like `EVP_MD_CTX_get_set_param_types`, and even worse, `EVP_MD_CTX_get_get_param_types`. I find that unacceptable. If you have a better idea, I'm all ears.

mattcaswell replied:
I'd prefer `EVP_MD_CTX_get_settable_types` and `EVP_MD_CTX_get_gettable_types`. But I do think we cannot just call it `EVP_MD_CTX_set_param_types`

mattcaswell replied:
Or another option: `EVP_MD_CTX_get_types_set_params`

levitte replied:
Settable and gettable types... Yes, I'm buying! Will do the change accross the board this even (that includes #8877)

paulidale replied:
The get/set nomenclature is confusing.  params.h uses them in one way and this in the opposite.  I agree with settable and gettable here and still want to rename the params.h functions to avoid misunderstandings.

t8m replied:
It would be even possible to drop the get - i.e. EVP_MD_CTX_settable_types() EVP_MD_CTX_gettable_types() - as there will be no set function, the get is almost redundant.

levitte replied:
Done

mattcaswell commented on doc/man3/EVP_EncryptInit.pod:
```diff
@@ -240,6 +253,27 @@ decrypting. If the B<pad> parameter is zero then no padding is
+B<params> from CIPHER context B<ctx>.
+
+EVP_CIPHER_get_param_types(), EVP_CIPHER_CTX_get_param_types(), and
+EVP_CIPHER_CTX_set_param_types() get a constant B<OSSL_PARAM> array
```
As above we have a "set" function which actually "gets"

mattcaswell commented on doc/man3/EVP_EncryptInit.pod:
```diff
@@ -240,6 +253,27 @@ decrypting. If the B<pad> parameter is zero then no padding is
+Get a B<OSSL_PARAM> array that describes the retrievable and settable
+parameters, i.e. parameters that can be used with EVP_MD_CTX_get_params()
+and EVP_MD_CTX_set_params(), respectively.
+See L<OSSL_PARAM(3)> for the use of B<OSSL_PARAM> as parameter descriptor.
```
I don't think you intended this paragraph to be here.

levitte replied:
I did, but it's not necessary any more...

mattcaswell commented:
Travis failure is relevant.

levitte committed:
squash! Add missing EVP param utility functions

RENAMED:

EVP_MD_get_param_types          EVP_MD_gettable_types
EVP_MD_CTX_set_param_types      EVP_MD_CTX_settable_types
EVP_MD_CTX_get_param_types      EVP_MD_CTX_gettable_types
EVP_CIPHER_get_param_types      EVP_CIPHER_gettable_types
EVP_CIPHER_CTX_set_param_types  EVP_CIPHER_CTX_settable_types
EVP_CIPHER_CTX_get_param_types  EVP_CIPHER_CTX_gettable_types

levitte committed:
squash! Modify 'openssl list' to give more verbose descriptions

REMEMBER RENAMES

levitte committed:
fixup! Add missing EVP param utility functions

levitte committed:
squash! Add {get,set}_param_types() functions for provider digests

REMEMBER RENAMES

levitte committed:
fixup! Add missing EVP param utility functions

levitte commented:
I hope to have answered all comments.

slontis commented on apps/list.c:
```diff
@@ -16,8 +16,94 @@
+        printed_len = BIO_snprintf(buf, bufsz, " [%d]", param->data_type);
+        if (printed_len > 0) {
+            buf += printed_len;
+            bufsz -= printed_len;
```
Maybe check bufsize is >= 0 (a bit late). Shouldnt happen - but it wont hurt..

levitte replied:
I trust BIO_snprintf to never go beyond the limit...

slontis commented on apps/list.c:
```diff
@@ -513,6 +616,23 @@ int list_main(int argc, char **argv)
+        unsigned int missing_help:1;
+        unsigned int objects:1;
+        unsigned int options:1;
+    } todo = { 0, };
```
Not overally impressed with this variable name :)

levitte replied:
Feel free to suggest something else

slontis replied:
flags, features ?

levitte replied:
[ahem] I'm not terribly impressed with those names :wink:

slontis replied:
todo3.0 ?

levitte replied:
Er...  you jest, I'm sure :wink:

slontis commented on crypto/evp/evp_enc.c:
```diff
@@ -1051,6 +1051,48 @@ int EVP_CIPHER_CTX_ctrl(EVP_CIPHER_CTX *ctx, int type, int arg, void *ptr)
+
+const OSSL_PARAM *EVP_CIPHER_gettable_types(const EVP_CIPHER *cipher)
+{
+    if (cipher != NULL && cipher->gettable_types != NULL)
```
In my PR that overlaps this one - I have used the name that was used in the MAC's. We need to decide which naming convention you want and apply it to all of them..

levitte replied:
Er, the name change is entirely driven by what *others* want, FYI.

slontis commented on providers/common/digests/sha3_prov.c:
```diff
@@ -228,7 +242,11 @@ const OSSL_DISPATCH name##_functions[] = {                              \
     { OSSL_FUNC_DIGEST_FREECTX, (void (*)(void))keccak_freectx },       \
     { OSSL_FUNC_DIGEST_DUPCTX, (void (*)(void))keccak_dupctx },         \
     { OSSL_FUNC_DIGEST_GET_PARAMS, (void (*)(void))name##_get_params }, \
+    { OSSL_FUNC_DIGEST_GETTABLE_TYPES,                                  \
```
Not sure there is a great correlation between the names OSSL_FUNC_DIGEST_GET_PARAMS & 
OSSL_FUNC_DIGEST_GETTABLE_TYPES. This is the type of the params that can be used by OSSL_FUNC_DIGEST_GET_PARAMS. It should at least have param in the name.

slontis commented on providers/common/include/internal/core_mkdigest.h:
```diff
@@ -42,6 +42,19 @@ static void *name##_dupctx(void *ctx) \
 static OSSL_OP_digest_get_params_fn name##_get_params;                  \
+static OSSL_OP_digest_gettable_types_fn name##_gettable_types;          \
+static const OSSL_PARAM known_##name##_gettable_types[] = {             \
+    {OSSL_DIGEST_PARAM_BLOCK_SIZE, OSSL_PARAM_INTEGER,                  \
```
Why are you macroing a const block?. This can be shared by all without duplicating it.

levitte replied:
This whole file could do with some refactoring, but that's for another PR

levitte committed:
squash! Add missing EVP param utility functions

RENAMED again:

EVP_MD_get_param_types          EVP_MD_gettable_params
EVP_MD_CTX_set_param_types      EVP_MD_CTX_settable_params
EVP_MD_CTX_get_param_types      EVP_MD_CTX_gettable_params
EVP_CIPHER_get_param_types      EVP_CIPHER_gettable_params
EVP_CIPHER_CTX_set_param_types  EVP_CIPHER_CTX_settable_params
EVP_CIPHER_CTX_get_param_types  EVP_CIPHER_CTX_gettable_params

levitte committed:
fixup! squash! Modify 'openssl list' to give more verbose descriptions

levitte committed:
fixup! Add {get,set}_param_types() functions for provider digests

levitte committed:
fixup! Add missing EVP param utility functions

levitte committed:
fixup! Add {get,set}_param_types() functions for provider digests

levitte committed:
fixup! Add missing EVP param utility functions

levitte commented:
In #9591 as well as here, the user API name pattern SOMETHING_{get,set}table_types was disputed as a bit disconnected from the pattern established with SOMETHING_{get,set}_params, so another name change happened and is now SOMETHING_{get,set}table_params.  Dispatch names are change to match.

If this is agreeable enough, we have a go (@slontis has approved #9591, but remarked that agreement on the name is needed)

levitte commented:
Something to be noted is that there's a slight disconnect in the dispatch function names for operation contexts.  The get,set functions are named `yadayada_ctx_{get,set}_params` while the gettable,settable functions are named `yadayada_{get,set}table_ctx_params`.

This could resolved in one of two ways:

1.  Rename all `yadayada_{get,set}table_ctx_params` to `yadayada_ctx_{get,set}table_params`
2.  Rename all `yadayada_ctx_{get,set}_params` to `yadayada_{get,set}_ctx_params`

I would prefer the 2nd alternative.  My reasoning is that in natural language, you would say "to get/set context parameters", not "context get/set parameters", making the former more intuitive.
We can't do much about the public API, that's governed by type names in a different manner.

levitte committed:
fixup! Add missing EVP param utility functions

slontis reviewed (approved):
Looks good to me - if others are happy with the names.

mattcaswell reviewed (approved):
LGTM.

With regards to the naming disconnect. I don't have a strong view, but perhaps a slight preference towards option 2. That doesn't need to be done in this PR though.

levitte commented:
> That doesn't need to be done in this PR though.

Good point.  I can prepare a separate PR for that rename.

levitte commented:
Merged.

ae3ff60e7bea6fb7510b5c0c2b9599d8430cf001 Add missing EVP param utility functions
ad623ec0cb2b2578a369431a8c3675b8bb2cc3b5 Modify 'openssl list' to give more verbose descriptions
ec02412b542d1240ed0ecf02e59ca4a4fa08027b Add {get,set}table_params() functions for provider digests

levitte commented:
Next up, #9591!
